### PR TITLE
Add predictor class

### PR DESCRIPTION
**Pull request by gitttt-1234**
## Summary by CodeRabbit

- **New Features**
	- Introduced new inference pipeline classes and functions for running pose estimation tasks in SLEAP-NN.
	- Added support for handling a maximum number of instances in data providers with a default set to 30.
- **Improvements**
	- Enhanced padding calculation in neural network architectures to support integers or lists.
	- Refined input shape handling in `SimpleUpsamplingBlock` for better accuracy.
	- Simplified the calculation of confidence maps for efficiency.
	- Adjusted bounding box offset calculation and iteration logic for improved centroid handling.
- **Documentation**
	- Added and adjusted empty lines after module docstrings for better readability.
- **Tests**
	- Expanded testing capabilities, including new fixtures and functionality for inference model testing.

### CODE REVIEW

File: sleap_nn/inference/inference.py [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+        if make_labels:
+            # Create SLEAP data structures from the predictions.
+            pred_labels = self._make_labeled_frames_from_generator(generator)
+            if save_labels:
+                sio.io.slp.write_labels(save_path, pred_labels)
+            return pred_labels
+
+        else:
+            # Just return the raw results.
+            return list(generator)
```
[the remark this answers is not in the document]
Comment (talmo) [reply sleap_nn/inference/inference.py:165]: Agreed. Let's just make it implicit and assume `save_labels = True` if `save_path is not None` so that we don't have to specify two separate arguments.

File: sleap_nn/inference/inference.py [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+
+    This model encapsulates the top-down approach where instances are first detected by
+    local peak detection of an anchor point and then cropped. These instance-centered
+    crops are then passed to an instance peak detector which is trained to detect all
+    remaining body parts for the instance that is centered within the crop.
+
+    Attributes:
+        centroid_crop: A centroid cropping layer. This can be either `CentroidCrop` or
+            `None`. This layer takes the full image as input and
+            outputs a set of centroids and cropped boxes. If `None`, the centroids are calculated with the provided anchor index using InstanceCentroid module and the centroid vals are set as 1.
```
Comment (talmo) [inline sleap_nn/inference/inference.py:279]: Newline at 88 characters

File: sleap_nn/inference/inference.py [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+        instance_peaks: A instance peak detection layer. This can be either
+            `FindInstancePeaks` or `None`. This layer takes as
+            input the output of the centroid cropper (if CentroidCrop not None else the image is cropped with the InstanceCropper module) and outputs the detected peaks for
```
Comment (talmo) [inline sleap_nn/inference/inference.py:282]: Newline at 88 characters

File: sleap_nn/inference/inference.py [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+            pass
+
+        if self.instance_peaks is None:
+            if "instance" in batch:
+                pass
+            else:
+                raise ValueError(
+                    "Ground truth data was not detected... "
+                    "Please load both models when predicting on non-ground-truth data."
+                )
```
Comment (talmo) [inline sleap_nn/inference/inference.py:329]: Add tests for this since it's a common use case

File: sleap_nn/inference/inference.py [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+            `"pred_instance_peaks": (batch_size, n_nodes, 2)`: Instance skeleton
+                points.
+            `"pred_peak_vals": (batch_size, n_nodes)`: Confidence
+                values for the instance skeleton points.
+        """
+        if self.centroid_crop is None:
+            batch["centroid_val"] = torch.ones(batch["instance"].shape[0])
+
+        else:
+            pass
```
Comment (talmo) [inline sleap_nn/inference/inference.py:320]: Why include this?

File: sleap_nn/inference/inference.py [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+            centroid_crop_layer = None
+        else:
+            self.model_config["centroid"] = self.centroid_config
+            self.model_config["data"] = self.centroid_config.inference_config.data
+            self.model_config["skeletons"] = self.centroid_config.data_config.skeletons
+            pass
+
+        # Create an instance of FindInstancePeaks layer if confmap_config is not None
+        if self.confmap_config is None:
+            pass
```
Comment (talmo) [inline sleap_nn/inference/inference.py:372]: Is this missing the logic described in the comment?

File: tests/assets/minimal_instance/last.ckpt
Comment (talmo) [inline tests/assets/minimal_instance/last.ckpt:1]: A 30 MB checkpoint is way too big to put in git!! It'll take longer and longer to clone the repo if we have several of these. Generate one that is actually _minimal_ by cranking down all the settings (filters, etc).

**Review by talmo — CHANGES_REQUESTED** [review]
